vowpalwabbit/learner.h

```diff
@@ -465,9 +515,51 @@ VW_WARNING_STATE_POP
     ret.learn_fd.multipredict_f = nullptr;
     ret.pred_type = pred_type;
     ret.is_multiline = std::is_same<multi_ex, E>::value;
+    ret._hash_index = std::type_index(typeid(T));
 
     return ret;
   }
+
+  std::type_index hash_index() { return _hash_index; }
+
+  inline VW::LEARNER::base_learner* get_base_reduction() { return learn_fd.base; }
+
+  void apply_from(
+    const VW::LEARNER::base_learner* base,
+    const std::unordered_map<std::type_index, VW::LEARNER::base_learner*>& reduction_template_map) {
+    if (base == nullptr || get_base_reduction() == nullptr) {
+      // currently working on base type. Nothing to do
+      return;
+    }
+    auto src = *this;
```

Comment:
Copy intentional? If so can you add a comment

Reply:
It is intentional, I need to keep a clean copy in order to re-write bits of it further down